hermeto/core/package_managers/pip/package_distributions.py

```diff
@@ -203,44 +235,149 @@ def process_package_distributions(
+        self.arch = self._parse_filter_spec(filters.arch)
+        self.os = self._parse_filter_spec(filters.os)
+        self.py_version = self._parse_filter_spec(filters.py_version)
+        self.py_impl = self._parse_filter_spec(filters.py_impl)
+
+    def __contains__(self, item: pypi_simple.DistributionPackage) -> bool:
+        """Check if the wheel matches the filter constraints."""
+        try:
+            _, _, _, tags = parse_wheel_filename(item.filename)
+        except InvalidWheelFilename:
+            log.warning("Skipping invalid wheel filename: %s", item.filename)
+            return False
+
+        # usually `tags` contain only one item
+        return any(self.matches(tag) for tag in tags)
+
+    def get_max_py_version(self) -> str:
+        """Get the maximum python version from the filter constraints."""
+        return max(self.py_version) if self.py_version is not None else ""
+
+    def parse_py_impl_and_py_version(self, interpreter: str) -> tuple[str, str]:
+        """
+        Examples:
+        >>> parse_py_impl_and_py_version("cp310")
+        ("cp", "310")
+        >>> parse_py_impl_and_py_version("py310")
+        ("py", "310")
+        >>> parse_py_impl_and_py_version("py3")
+        ("py", "3")
+        """
+        match = re.fullmatch(r"([a-z]+)(\d+)", interpreter)
+        if match is None:
+            return "", ""
+
+        return match.group(1), match.group(2)
+
+    def matches(self, tag: Tag) -> bool:
+        """
+        Check if the filter constraints match the platform compatibility tag from the wheel filename.
+
+        - multiple values in a field are combined with OR logic
+        - multiple fields are combined with AND logic
+        - if a field is not provided (None), it is treated as `:all:` and treated as a match
+
+        See https://packaging.pypa.io/en/stable/tags.html
+        """
+        valid_arch = (
+            self.arch is None
+            or tag.platform == "any"
+            or any(arch in tag.platform for arch in self.arch)
+        )
+        # check operating system compatibility
+        valid_os = (
+            self.os is None or tag.platform == "any" or any(os in tag.platform for os in self.os)
+        )
+        # check Python implementation compatibility
+        valid_py_impl = (
+            self.py_impl is None
+            or "py" in tag.interpreter
+            or any(impl in tag.interpreter for impl in self.py_impl)
+        )
+
+        # check Python version compatibility with special handling for generic vs specific tags
+        max_py_version = self.get_max_py_version()
+        wheel_py_impl, wheel_py_version = self.parse_py_impl_and_py_version(tag.interpreter)
+
+        if wheel_py_impl == "py":
+            # generic "py" tags are compatible if version <= max_version and abi allows it
+            valid_py_version = self.py_version is None or (
+                wheel_py_version <= max_py_version and tag.abi in ("abi3", "none")
```

Comment:
I think this is incorrect assessment - `abi3` is shorthand for `cp3\d` and so if python tag is `py*`, abi is irrelevant, it's automatically matching, abi handling needs to go the `else` branch and iff the implementation user requested is `cp*`, otherwise only `none` can be used as the ever-matching default value along with `cp*`. I'm wondering if we should let users specify also the ABI tag.

Reply:
No, IMHO if we can all agree on this alternative approach to map the input fields to those compatibility tags, then `os` and `architecture` would have to become `platform` otherwise it's confusing. And that is even at the cost of our binary filter fields not being consistent across the backend range which can support binary filtering, because that's what's currently kind of happening, we're trying to keep the common input fields consistent and add some more for Python where `os` and `architecture` isn't enough.